torchao/csrc/metal/int4_quantized_kernels.metal

```diff
@@ -0,0 +1,128 @@
+#include <metal_simdgroup>
+#include <metal_stdlib>
+using namespace metal;
+
+template <typename T> struct Vec4Type {};
+
+template <> struct Vec4Type<float> { using type = float4; };
+
+template <> struct Vec4Type<half> { using type = half4; };
+
+#if __METAL_VERSION__ >= 310
+template <> struct Vec4Type<bfloat> { using type = bfloat4; };
+#endif
+
+template <typename T, unsigned groupSize>
+kernel void int4pack_mv(constant T *A [[buffer(0)]],
+                        constant uchar *B [[buffer(1)]],
+                        constant T *scalesAndZeros [[buffer(2)]],
+                        device T *outputData [[buffer(3)]],
+                        constant uint3 &sizes [[buffer(4)]], // M, K, N
+                        uint thread_index [[thread_position_in_grid]],
+                        uint tid_in_simdgroup [[thread_index_in_simdgroup]]) {
+  constexpr uint threads_per_channel = 32;
+  constexpr uint ks_per_thread = 4;
+  constexpr uint k_pack_factor = 2;
+  const uint K = sizes.y;
+  const uint N = sizes.z;
+  uint n = thread_index; // 0..N/4-1
+  n = n / threads_per_channel;
+  n = n * 4;
+  uint k = (tid_in_simdgroup % threads_per_channel) * ks_per_thread;
+  constexpr int k_jump = threads_per_channel * ks_per_thread;
```

Comment:
would like to see another explanatory comment on the algorithm here so that I don't have to deeply understand almost every line of the implementation (and probably draw my own picture) just to figure out the algorithm in use. make it clear and easy! could use hints on
- the strategy for mapping threads to the matrix dimension
- what each thread is responsible for (looks like it iterates along the k dimension, so it handles some fixed amount of entire columns of the matrix?)